Algorithm.Framework/Risk/MaximumDrawdownPercentPortfolio.cs

```diff
@@ -0,0 +1,71 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+using System;
+using System.Collections.Generic;
+using QuantConnect.Algorithm.Framework.Portfolio;
+using QuantConnect.Securities;
+
+namespace QuantConnect.Algorithm.Framework.Risk
+{
+    /// <summary>
+    /// Provides an implementation of <see cref="IRiskManagementModel"/> that limits the drawdown
+    /// of the portfolio to the specified percentage
+    /// </summary>
+    public class MaximumDrawdownPercentPortfolio : RiskManagementModel
+    {
+        private readonly decimal _maximumDrawdownPercent;
+        private decimal _startingValue;
+        private bool _initialised = false;
+
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MaximumDrawdownPercentPortfolio"/> class
+        /// </summary>
+        /// <param name="maximumDrawdownPercent">The maximum percentage drawdown allowed for algorithm portfolio
+        /// compared with starting value, defaults to 5% drawdown</param>
+        public MaximumDrawdownPercentPortfolio(decimal maximumDrawdownPercent = 0.05m)
+        {
+            _maximumDrawdownPercent = -Math.Abs(maximumDrawdownPercent);
+        }
+
+        /// <summary>
+        /// Manages the algorithm's risk at each time step
+        /// </summary>
+        /// <param name="algorithm">The algorithm instance</param>
+        /// <param name="targets">The current portfolio targets to be assessed for risk</param>
+        public override IEnumerable<IPortfolioTarget> ManageRisk(QCAlgorithmFramework algorithm, IPortfolioTarget[] targets)
+        {
+            if(!_initialised)
+            {
+                _startingValue = algorithm.Portfolio.TotalPortfolioValue; // Set inital portfolio value
+                _initialised = true;
+            }
+
+            var pnl = GetTotalDrawdownPercent(algorithm.Portfolio.TotalPortfolioValue);
+            if (pnl < _maximumDrawdownPercent)
+            {
+                foreach(var target in targets)
+                    yield return new PortfolioTarget(target.Symbol, 0);
+            }
+        }
+
+        private decimal GetTotalDrawdownPercent(decimal currentValue)
+        {
+            return (currentValue / _startingValue) - 1.0m;
```

Comment:
This appears to be a 'global' drawdown, meaning from the time the algorithm first launched. Do we need a way to 'reset' the `currentValue` ? For example, consider if an algorithm runs for a year it may be up 100% w/ this model set to a 5% max drawdown and then from there drop 50% and this check still wouldn't trigger a liquidate -- is that intended? If this _is_ intended, perhaps some extra documentation describing that.

I think the ideal here would be to change `currentValue` to be maximum value, so on each iteration check if `TotalPortfolioValue` is greater than the previous and reset it. We could also set it each time we perform a global liquidate to set it to the new lower threshold value.

thoughts?

Reply:
Yes, I like that implementation. To me it sounds like a Trailing Stop Loss rather than an Absolute Stop Loss. Perhaps a user would use both Risk Management Models - with the Absolute Stop Loss set to a lower drawdown than the Trailing? If so, I could rename this implementation to make it clear it's an Absolute Stop Loss rather than Trailing - then implement Trailing at a later time?

Reply:
Naming and documentation are key! So long as it's properly named/documented I'm happy with whichever route you choose 👍

Reply:
@mchandschuh I've added a constructor flag so both options can be used and documented it. Is this fine or should it really be two separate classes?